### PR TITLE
Add "Extending DIALS" page to docs

### PR DESCRIPTION
**Pull request by rjgildea**
Start of a developer documentation page on how to extend DIALS.

Currently addresses:
- adding dxtbx.format classes
- adding dials.index basis_vector_search strategies

TODO:
- [x] adding dials.index lattice search strategies
- [ ] adding profile models
- [x] adding scaling models

### CODE REVIEW

File: doc/sphinx/documentation/developer_examples/extending_dials.rst [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+      {
+          "dxtbx.format": [
+              "FormatCBFMiniPilatusMyBeamline:FormatCBFMiniPilatus = myproject.my_format_module:FormatCBFMiniPilatusMyBeamline",
+          ],
+      }
+  )
+
+More generally, the format of an entry point for dxtbx.format is::
+
+    "FormatMyClass:FormatBaseClass1,FormatBaseClass2 = myproject.my_format_module:my"
```
Comment (benjaminhwilliams) [inline doc/sphinx/documentation/developer_examples/extending_dials.rst]: `... = myproject.my_format_module:MyFormatClass`?

Or am I reading this incorrectly?

File: doc/sphinx/documentation/developer_examples/extending_dials.rst [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  mystrategy_phil_str = """\
+  magic_parameter = 42
+      .help = "This is a magic parameter."
+      .type = float
+  """
+
+  class MyStrategy(Strategy):
+      """Basis vector search using my magic algorithm."""
+
+  phil_scope = phil.parse(mystrategy_phil_str)
```
Comment (benjaminhwilliams) [inline doc/sphinx/documentation/developer_examples/extending_dials.rst]: Indentation.

File: doc/sphinx/documentation/developer_examples/extending_dials.rst [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+  def find_basis_vectors(self, reciprocal_lattice_vectors):
```
Comment (benjaminhwilliams) [inline doc/sphinx/documentation/developer_examples/extending_dials.rst]: Indentation.

File: doc/sphinx/documentation/developer_examples/extending_dials.rst (hunk 1 above)
Comment (rjgildea) [reply doc/sphinx/documentation/developer_examples/extending_dials.rst]: Not sure what happened there, now fixed.

Comment (benjaminhwilliams) [reply doc/sphinx/documentation/developer_examples/extending_dials.rst]: Cool.  You successfully unmangled my mangled suggestion!  :+1:

**Comment by jbeilstenedmands** [comment]
I'm working on the scaling models section, but turns out my original implementation is stupid so refactoring the code first.

**Comment by graeme-winter** [comment]
> I'm working on the scaling models section, but turns out my original implementation is stupid so refactoring the code first.

One of the reasons why documentation is good - a bit like soft toy debugging :)

**Pull request merged**